src/modules/krita/producer_krita.c

```diff
@@ -0,0 +1,231 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include <framework/mlt.h>
+#include <limits.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <framework/mlt_factory.h>
+#include <framework/mlt_frame.h>
+#include <framework/mlt_producer.h>
+#include <framework/mlt_property.h>
+#include <framework/mlt_service.h>
+
+typedef struct
+{
+    int frame_start;
+    int frame_end;
+    int limit_enabled;
+    double speed;
+    mlt_producer producer_internal;
+} private_data;
+
+static int restrict_range(int input, int min, int max)
+{
+    const int span = max - min;
+    return (MAX(input - min, 0) % (span + 1)) + min;
+}
+
+int is_valid_range(const private_data *pdata)
+{
+    return pdata->frame_start > -1 && pdata->frame_end > -1
+           && pdata->frame_end > pdata->frame_start;
+}
+
+int is_limit_enabled(const private_data *pdata)
+{
+    return pdata->limit_enabled != 0;
+}
+
+static int producer_get_audio(mlt_frame frame,
+                              void **buffer,
+                              mlt_audio_format *format,
+                              int *frequency,
+                              int *channels,
+                              int *samples)
+{
+    mlt_producer producer = mlt_frame_pop_audio(frame);
+    private_data *pdata = (private_data *) producer->child;
+    struct mlt_audio_s audio;
+    mlt_audio_set_values(&audio, *buffer, *frequency, *format, *samples, *channels);
+
+    int error = mlt_frame_get_audio(frame,
+                                    &audio.data,
+                                    &audio.format,
+                                    &audio.frequency,
+                                    &audio.channels,
+                                    &audio.samples);
+
+    // Scale the frequency to account for the speed change.
+    audio.frequency = (double) audio.frequency * fabs(pdata->speed);
+
+    if (pdata->speed < 0.0) {
+        mlt_audio_reverse(&audio);
+    }
+
+    mlt_audio_get_values(&audio, buffer, frequency, format, samples, channels);
+
+    return error;
+}
+
+static int producer_get_frame(mlt_producer producer, mlt_frame_ptr frame, int index)
+{
+    private_data *pdata = (private_data *) producer->child;
+
+    const int position = mlt_producer_position(pdata->producer_internal);
+    if (is_valid_range(pdata) && is_limit_enabled(pdata)) {
+        mlt_properties_set_position(MLT_PRODUCER_PROPERTIES(pdata->producer_internal),
```

Comment:
Does mlt_producer_seek() work here instead of modifying a private property directly?